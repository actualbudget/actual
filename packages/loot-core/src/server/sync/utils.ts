export function isError(value: unknown): value is { error: unknown } {
  return (value as { error: unknown }).error !== undefined;
}
