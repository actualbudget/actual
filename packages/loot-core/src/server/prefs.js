import * as fs from '../platform/server/fs';

import { Timestamp } from './crdt';
import { sendMessages } from './sync';

let prefs = null;

export async function loadPrefs(id) {
  if (process.env.NODE_ENV === 'test' && !id) {
    prefs = { dummyTestPrefs: true };
    return prefs;
  }

  const fullpath = fs.join(fs.getBudgetDir(id), 'metadata.json');

  try {
    prefs = JSON.parse(await fs.readFile(fullpath));
  } catch (e) {
    // If the user messed something up, be flexible and allow them to
    // still load the budget database. Default the budget name to the
    // id.
    prefs = { id, budgetName: id };
  }

  // No matter what is in `id` field, force it to be the current id.
  // This makes it resilient to users moving around folders, etc
  prefs.id = id;
  return prefs;
}

export async function savePrefs(prefsToSet, { avoidSync = false } = {}) {
  Object.assign(prefs, prefsToSet);

  if (!avoidSync) {
    // Sync whitelisted prefs
    let messages = Object.keys(prefsToSet)
      .map(key => {
        if (key === 'budgetType' || key === 'budgetName') {
          return {
            dataset: 'prefs',
            row: key,
            column: 'value',
            value: prefsToSet[key],
            timestamp: Timestamp.send(),
          };
        }
        return null;
      })
      .filter(x => x);

    if (messages.length > 0) {
      await sendMessages(messages);
    }
  }

  if (!prefs.dummyTestPrefs) {
    let prefsPath = fs.join(fs.getBudgetDir(prefs.id), 'metadata.json');
    await fs.writeFile(prefsPath, JSON.stringify(prefs));
  }
}

export function unloadPrefs() {
  prefs = null;
}

export function getPrefs() {
  return prefs;
}

export function getDefaultPrefs(id, budgetName) {
  // Add any notifications in here that new users shouldn't see.
  // Without them, a popup will show to explain a new feature.
  return {
    id,
    budgetName,
    'notifications.schedules': true,
    'notifications.repair-splits': true,
  };
}

export async function readPrefs(id) {
  const fullpath = fs.join(fs.getBudgetDir(id), 'metadata.json');

  try {
    return JSON.parse(await fs.readFile(fullpath));
  } catch (e) {
    return null;
  }
}
