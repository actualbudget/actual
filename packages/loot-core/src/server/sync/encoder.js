import * as encryption from '../encryption';
import { SyncError } from '../errors';
import * as prefs from '../prefs';

import * as SyncPb from './proto/sync_pb';

function coerceBuffer(value) {
  // The web encryption APIs give us back raw Uint8Array... but our
  // encryption code assumes we can work with it as a buffer. This is
  // a leaky abstraction and ideally the our abstraction over the web
  // encryption APIs should do this.
  if (!Buffer.isBuffer(value)) {
    return Buffer.from(value);
  }
  return value;
}

export async function encode(groupId, fileId, since, messages) {
  let { encryptKeyId } = prefs.getPrefs();
  let requestPb = new SyncPb.SyncRequest();

  for (let i = 0; i < messages.length; i++) {
    let msg = messages[i];
    let envelopePb = new SyncPb.MessageEnvelope();
    envelopePb.setTimestamp(msg.timestamp);

    let messagePb = new SyncPb.Message();
    messagePb.setDataset(msg.dataset);
    messagePb.setRow(msg.row);
    messagePb.setColumn(msg.column);
    messagePb.setValue(msg.value);
    let binaryMsg = messagePb.serializeBinary();

    if (encryptKeyId) {
      let encrypted = new SyncPb.EncryptedData();

      let result;
      try {
        result = await encryption.encrypt(binaryMsg, encryptKeyId);
      } catch (e) {
        throw new SyncError('encrypt-failure', {
          isMissingKey: e.message === 'missing-key',
        });
      }

      encrypted.setData(result.value);
      encrypted.setIv(Buffer.from(result.meta.iv, 'base64'));
      encrypted.setAuthtag(Buffer.from(result.meta.authTag, 'base64'));

      envelopePb.setContent(encrypted.serializeBinary());
      envelopePb.setIsencrypted(true);
    } else {
      envelopePb.setContent(binaryMsg);
    }

    requestPb.addMessages(envelopePb);
  }

  requestPb.setGroupid(groupId);
  requestPb.setFileid(fileId);
  requestPb.setKeyid(encryptKeyId);
  requestPb.setSince(since);

  return requestPb.serializeBinary();
}

export async function decode(data) {
  let { encryptKeyId } = prefs.getPrefs();

  let responsePb = SyncPb.SyncResponse.deserializeBinary(data);
  let merkle = JSON.parse(responsePb.getMerkle());
  let list = responsePb.getMessagesList();
  let messages = [];

  for (let i = 0; i < list.length; i++) {
    let envelopePb = list[i];
    let timestamp = envelopePb.getTimestamp();
    let encrypted = envelopePb.getIsencrypted();
    let msg;

    if (encrypted) {
      let binary = SyncPb.EncryptedData.deserializeBinary(
        envelopePb.getContent(),
      );

      let decrypted;
      try {
        decrypted = await encryption.decrypt(coerceBuffer(binary.getData()), {
          keyId: encryptKeyId,
          algorithm: 'aes-256-gcm',
          iv: coerceBuffer(binary.getIv()),
          authTag: coerceBuffer(binary.getAuthtag()),
        });
      } catch (e) {
        console.log(e);
        throw new SyncError('decrypt-failure', {
          isMissingKey: e.message === 'missing-key',
        });
      }

      msg = SyncPb.Message.deserializeBinary(decrypted);
    } else {
      msg = SyncPb.Message.deserializeBinary(envelopePb.getContent());
    }

    messages.push({
      timestamp: timestamp,
      dataset: msg.getDataset(),
      row: msg.getRow(),
      column: msg.getColumn(),
      value: msg.getValue(),
    });
  }

  return { messages, merkle };
}
