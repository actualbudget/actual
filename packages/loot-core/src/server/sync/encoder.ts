import { Timestamp, SyncProtoBuf } from '@actual-app/crdt';

import * as encryption from '../encryption';
import { SyncError } from '../errors';
import * as prefs from '../prefs';

import { Message } from './index';

function coerceBuffer(value) {
  // The web encryption APIs give us back raw Uint8Array... but our
  // encryption code assumes we can work with it as a buffer. This is
  // a leaky abstraction and ideally the our abstraction over the web
  // encryption APIs should do this.
  if (!Buffer.isBuffer(value)) {
    return Buffer.from(value);
  }
  return value;
}

export async function encode(
  groupId: string,
  fileId: string,
  since: Timestamp | string,
  messages: Message[],
): Promise<Uint8Array> {
  let { encryptKeyId } = prefs.getPrefs();
  let requestPb = new SyncProtoBuf.SyncRequest();

  for (let i = 0; i < messages.length; i++) {
    let msg = messages[i];
    let envelopePb = new SyncProtoBuf.MessageEnvelope();
    envelopePb.setTimestamp(msg.timestamp.toString());

    let messagePb = new SyncProtoBuf.Message();
    messagePb.setDataset(msg.dataset);
    messagePb.setRow(msg.row);
    messagePb.setColumn(msg.column);
    messagePb.setValue(msg.value as string);
    let binaryMsg = messagePb.serializeBinary();

    if (encryptKeyId) {
      let encrypted = new SyncProtoBuf.EncryptedData();

      let result;
      try {
        result = await encryption.encrypt(binaryMsg, encryptKeyId);
      } catch (e) {
        throw new SyncError('encrypt-failure', {
          isMissingKey: e.message === 'missing-key',
        });
      }

      encrypted.setData(result.value);
      encrypted.setIv(Buffer.from(result.meta.iv, 'base64'));
      encrypted.setAuthtag(Buffer.from(result.meta.authTag, 'base64'));

      envelopePb.setContent(encrypted.serializeBinary());
      envelopePb.setIsencrypted(true);
    } else {
      envelopePb.setContent(binaryMsg);
    }

    requestPb.addMessages(envelopePb);
  }

  requestPb.setGroupid(groupId);
  requestPb.setFileid(fileId);
  requestPb.setKeyid(encryptKeyId);
  requestPb.setSince(since.toString());

  return requestPb.serializeBinary();
}

export async function decode(
  data: Uint8Array,
): Promise<{ messages: Message[]; merkle: { hash: number } }> {
  let { encryptKeyId } = prefs.getPrefs();

  let responsePb = SyncProtoBuf.SyncResponse.deserializeBinary(data);
  let merkle = JSON.parse(responsePb.getMerkle());
  let list = responsePb.getMessagesList();
  let messages = [];

  for (let i = 0; i < list.length; i++) {
    let envelopePb = list[i];
    let timestamp = Timestamp.parse(envelopePb.getTimestamp());
    let encrypted = envelopePb.getIsencrypted();
    let msg;

    if (encrypted) {
      let binary = SyncProtoBuf.EncryptedData.deserializeBinary(
        envelopePb.getContent() as Uint8Array,
      );

      let decrypted;
      try {
        decrypted = await encryption.decrypt(coerceBuffer(binary.getData()), {
          keyId: encryptKeyId,
          algorithm: 'aes-256-gcm',
          iv: coerceBuffer(binary.getIv()),
          authTag: coerceBuffer(binary.getAuthtag()),
        });
      } catch (e) {
        console.log(e);
        throw new SyncError('decrypt-failure', {
          isMissingKey: e.message === 'missing-key',
        });
      }

      msg = SyncProtoBuf.Message.deserializeBinary(decrypted);
    } else {
      msg = SyncProtoBuf.Message.deserializeBinary(
        envelopePb.getContent() as Uint8Array,
      );
    }

    messages.push({
      timestamp: timestamp,
      dataset: msg.getDataset(),
      row: msg.getRow(),
      column: msg.getColumn(),
      value: msg.getValue(),
    });
  }

  return { messages, merkle };
}
