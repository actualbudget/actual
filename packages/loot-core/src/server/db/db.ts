// @ts-strict-ignore
import {
  makeClock,
  setClock,
  serializeClock,
  deserializeClock,
  makeClientId,
  Timestamp,
} from '@actual-app/crdt';
import { Database } from '@jlongster/sql.js';
import { LRUCache } from 'lru-cache';
import { v4 as uuidv4 } from 'uuid';

import * as fs from '../../platform/server/fs';
import * as sqlite from '../../platform/server/sqlite';
import {
  schema,
  schemaConfig,
  convertForInsert,
  convertForUpdate,
  convertFromSelect,
} from '../aql';
import { sendMessages } from '../sync';

import { DbClockMessage } from './types';

export * from './types';

export { toDateRepr, fromDateRepr } from '../models';

let dbPath: string | null = null;
let db: Database | null = null;

// Util

export function getDatabasePath() {
  return dbPath;
}

export async function openDatabase(id?: string) {
  if (db) {
    await sqlite.closeDatabase(db);
  }

  dbPath = fs.join(fs.getBudgetDir(id), 'db.sqlite');
  setDatabase(await sqlite.openDatabase(dbPath));

  // await execQuery('PRAGMA journal_mode = WAL');
}

export async function closeDatabase() {
  if (db) {
    await sqlite.closeDatabase(db);
    setDatabase(null);
  }
}

export function setDatabase(db_: Database) {
  db = db_;
  resetQueryCache();
}

export function getDatabase() {
  return db;
}

export async function loadClock() {
  const row = await first<DbClockMessage>('SELECT * FROM messages_clock');
  if (row) {
    const clock = deserializeClock(row.clock);
    setClock(clock);
  } else {
    // No clock exists yet (first run of the app), so create a default
    // one.
    const timestamp = new Timestamp(0, 0, makeClientId());
    const clock = makeClock(timestamp);
    setClock(clock);

    await runQuery('INSERT INTO messages_clock (id, clock) VALUES (?, ?)', [
      1,
      serializeClock(clock),
    ]);
  }
}

// Functions
export function runQuery(
  sql: string,
  params?: Array<string | number>,
  fetchAll?: false,
): { changes: unknown };

export function runQuery<T>(
  sql: string,
  params: Array<string | number> | undefined,
  fetchAll: true,
): T[];

export function runQuery<T>(
  sql: string,
  params: (string | number)[],
  fetchAll: boolean,
) {
  if (fetchAll) {
    return sqlite.runQuery<T>(db, sql, params, true);
  } else {
    return sqlite.runQuery(db, sql, params, false);
  }
}

export function execQuery(sql: string) {
  sqlite.execQuery(db, sql);
}

// This manages an LRU cache of prepared query statements. This is
// only needed in hot spots when you are running lots of queries.
let _queryCache = new LRUCache<string, string>({ max: 100 });
export function cache(sql: string) {
  const cached = _queryCache.get(sql);
  if (cached) {
    return cached;
  }

  const prepared = sqlite.prepare(db, sql);
  _queryCache.set(sql, prepared);
  return prepared;
}

function resetQueryCache() {
  _queryCache = new LRUCache<string, string>({ max: 100 });
}

export function transaction(fn: () => void) {
  return sqlite.transaction(db, fn);
}

export function asyncTransaction(fn: () => Promise<void>) {
  return sqlite.asyncTransaction(db, fn);
}

// This function is marked as async because `runQuery` is no longer
// async. We return a promise here until we've audited all the code to
// make sure nothing calls `.then` on this.
export async function all<T>(sql: string, params?: (string | number)[]) {
  return runQuery<T>(sql, params, true);
}

export async function first<T>(sql, params?: (string | number)[]) {
  const arr = await runQuery<T>(sql, params, true);
  return arr.length === 0 ? null : arr[0];
}

// The underlying sql system is now sync, but we can't update `first` yet
// without auditing all uses of it
export function firstSync<T>(sql, params?: (string | number)[]) {
  const arr = runQuery<T>(sql, params, true);
  return arr.length === 0 ? null : arr[0];
}

// This function is marked as async because `runQuery` is no longer
// async. We return a promise here until we've audited all the code to
// make sure nothing calls `.then` on this.
export async function run(sql, params?: (string | number)[]) {
  return runQuery(sql, params);
}

export async function select(table, id) {
  const rows = await runQuery(
    'SELECT * FROM ' + table + ' WHERE id = ?',
    [id],
    true,
  );
  // TODO: In the next phase, we will make this function generic
  // and pass the type of the return type to `runQuery`.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return rows[0] as any;
}

export async function update(table, params) {
  const fields = Object.keys(params).filter(k => k !== 'id');

  if (params.id == null) {
    throw new Error('update: id is required');
  }

  await sendMessages(
    fields.map(k => {
      return {
        dataset: table,
        row: params.id,
        column: k,
        value: params[k],
        timestamp: Timestamp.send(),
      };
    }),
  );
}

export async function insertWithUUID(table, row) {
  if (!row.id) {
    row = { ...row, id: uuidv4() };
  }

  await insert(table, row);

  // We can't rely on the return value of insert because if the
  // primary key is text, sqlite returns the internal row id which we
  // don't care about. We want to return the generated UUID.
  return row.id;
}

export async function insert(table, row) {
  const fields = Object.keys(row).filter(k => k !== 'id');

  if (row.id == null) {
    throw new Error('insert: id is required');
  }

  await sendMessages(
    fields.map(k => {
      return {
        dataset: table,
        row: row.id,
        column: k,
        value: row[k],
        timestamp: Timestamp.send(),
      };
    }),
  );
}

export async function delete_(table, id) {
  await sendMessages([
    {
      dataset: table,
      row: id,
      column: 'tombstone',
      value: 1,
      timestamp: Timestamp.send(),
    },
  ]);
}

export async function deleteAll(table: string) {
  const rows = await all<{ id: string }>(`
    SELECT id FROM ${table} WHERE tombstone = 0
  `);
  await Promise.all(rows.map(({ id }) => delete_(table, id)));
}

export async function selectWithSchema(table, sql, params) {
  const rows = await runQuery(sql, params, true);
  const convertedRows = rows
    .map(row => convertFromSelect(schema, schemaConfig, table, row))
    .filter(Boolean);
  // TODO: Make convertFromSelect generic so we don't need this cast
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return convertedRows as any[];
}

export async function selectFirstWithSchema(table, sql, params) {
  const rows = await selectWithSchema(table, sql, params);
  return rows.length > 0 ? rows[0] : null;
}

export function insertWithSchema(table, row) {
  // Even though `insertWithUUID` does this, we need to do it here so
  // the schema validation passes
  if (!row.id) {
    row = { ...row, id: uuidv4() };
  }

  return insertWithUUID(
    table,
    convertForInsert(schema, schemaConfig, table, row),
  );
}

export function updateWithSchema(table, fields) {
  return update(table, convertForUpdate(schema, schemaConfig, table, fields));
}
