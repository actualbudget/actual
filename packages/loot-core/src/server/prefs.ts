import { Timestamp } from '@actual-app/crdt';

import * as fs from '../platform/server/fs';
import type { LocalPrefs } from '../types/prefs';

import { Message, sendMessages } from './sync';

export const BUDGET_TYPES = ['report', 'rollover'] as const;
export type BudgetType = (typeof BUDGET_TYPES)[number];

let prefs: LocalPrefs = null;

export async function loadPrefs(id?: string): Promise<LocalPrefs> {
  if (process.env.NODE_ENV === 'test' && !id) {
    prefs = getDefaultPrefs('test', 'test_LocalPrefs');
    return prefs;
  }

  const fullpath = fs.join(fs.getBudgetDir(id), 'metadata.json');

  try {
    prefs = JSON.parse(await fs.readFile(fullpath));
  } catch (e) {
    // If the user messed something up, be flexible and allow them to
    // still load the budget database. Default the budget name to the
    // id.
    prefs = { id, budgetName: id };
  }

  // delete released feature flags
  let releasedFeatures = ['syncAccount'];
  for (const feature of releasedFeatures) {
    delete prefs[`flags.${feature}`];
  }

  // delete legacy notifications
  for (const key of Object.keys(prefs)) {
    if (key.startsWith('notifications.')) {
      delete prefs[key];
    }
  }

  // No matter what is in `id` field, force it to be the current id.
  // This makes it resilient to users moving around folders, etc
  prefs.id = id;
  return prefs;
}

export async function savePrefs(
  prefsToSet: LocalPrefs,
  { avoidSync = false } = {},
): Promise<void> {
  Object.assign(prefs, prefsToSet);

  if (!avoidSync) {
    // Sync whitelisted prefs
    const messages: Message[] = Object.keys(prefsToSet)
      .map(key => {
        if (key === 'budgetType' || key === 'budgetName') {
          return {
            dataset: 'prefs',
            row: key,
            column: 'value',
            value: prefsToSet[key],
            timestamp: Timestamp.send(),
          };
        }
        return null;
      })
      .filter(x => x);

    if (messages.length > 0) {
      await sendMessages(messages);
    }
  }

  if (process.env.NODE_ENV !== 'test') {
    let prefsPath = fs.join(fs.getBudgetDir(prefs.id), 'metadata.json');
    await fs.writeFile(prefsPath, JSON.stringify(prefs));
  }
}

export function unloadPrefs(): void {
  prefs = null;
}

export function getPrefs(): LocalPrefs {
  return prefs;
}

export function getDefaultPrefs(id: string, budgetName: string): LocalPrefs {
  return { id, budgetName };
}
