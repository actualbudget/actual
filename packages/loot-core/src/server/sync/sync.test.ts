import { getClock, Timestamp } from '@actual-app/crdt';

import * as db from '../db';
import * as prefs from '../prefs';
import * as sheet from '../sheet';
import * as mockSyncServer from '../tests/mockSyncServer';

import * as encoder from './encoder';
import { isError } from './utils';

import { setSyncingMode, sendMessages, applyMessages, fullSync } from './index';

beforeEach(() => {
  mockSyncServer.reset();
  setSyncingMode('enabled');
  return global.emptyDatabase()();
});

afterEach(() => {
  global.resetTime();
  setSyncingMode('disabled');
});

describe('Sync', () => {
  it('should send messages to the server', async () => {
    prefs.loadPrefs();
    prefs.savePrefs({ groupId: 'group' });

    let timestamp = Timestamp.send();
    await sendMessages([
      {
        dataset: 'transactions',
        row: 'foo',
        column: 'amount',
        value: 3200,
        timestamp,
      },
    ]);

    global.stepForwardInTime();

    timestamp = Timestamp.send();
    await sendMessages([
      {
        dataset: 'transactions',
        row: 'foo',
        column: 'amount',
        value: 4200,
        timestamp,
      },
    ]);

    expect(getClock().timestamp.toString()).toEqual(timestamp.toString());
    expect(mockSyncServer.getClock().merkle).toEqual(getClock().merkle);

    expect(await db.all('SELECT * FROM messages_crdt')).toMatchSnapshot();
    expect(await db.all('SELECT * FROM messages_clock')).toMatchSnapshot();
  });

  it('should resend old messages to the server', async () => {
    prefs.loadPrefs();
    prefs.savePrefs({ groupId: 'group' });

    global.stepForwardInTime(Date.parse('2018-11-13T13:20:00.000Z'));

    await applyMessages([
      global.stepForwardInTime() || {
        dataset: 'transactions',
        row: 'foo',
        column: 'amount',
        value: 3200,
        timestamp: Timestamp.send(),
      },
      global.stepForwardInTime() || {
        dataset: 'transactions',
        row: 'foo',
        column: 'amount',
        value: 3200,
        timestamp: Timestamp.send(),
      },
    ]);

    // Move the clock forward so that the above 2 messages are not
    // automatically sent out, but will need to be re-sent by way of
    // the merkle tree
    prefs.savePrefs({ lastSyncedTimestamp: getClock().timestamp.toString() });

    expect(mockSyncServer.getMessages().length).toBe(0);

    const result = await fullSync();
    if (isError(result)) throw result.error;
    expect(result.messages.length).toBe(0);
    expect(mockSyncServer.getMessages().length).toBe(2);
  });

  it('should sync multiple clients', async () => {
    prefs.loadPrefs();
    prefs.savePrefs({
      groupId: 'group',
      lastSyncedTimestamp: Timestamp.zero.toString(),
    });

    await mockSyncServer.handlers['/sync/sync'](
      await encoder.encode(
        'group',
        'client',
        Timestamp.parse('1970-01-01T01:17:37.000Z-0000-0000testinguuid2'),
        [
          {
            dataset: 'transactions',
            row: 'foo',
            column: 'amount',
            value: 'N:3200',
            timestamp: Timestamp.parse(
              '1970-01-02T05:17:36.789Z-0000-0000testinguuid2',
            ),
          },
          {
            dataset: 'transactions',
            row: 'foo',
            column: 'amount',
            value: 'N:4200',
            timestamp: Timestamp.parse(
              '1970-01-02T10:17:36.999Z-0000-0000testinguuid2',
            ),
          },
        ],
      ),
    );

    await applyMessages([
      global.stepForwardInTime(Date.parse('1970-01-03T10:17:37.000Z')) || {
        dataset: 'transactions',
        row: 'foo',
        column: 'amount',
        value: 5000,
        timestamp: Timestamp.send(),
      },
    ]);

    const result = await fullSync();
    if (isError(result)) throw result.error;
    expect(result.messages.length).toBe(2);
    expect(mockSyncServer.getMessages().length).toBe(3);
  });
});

async function registerBudgetMonths(months) {
  let createdMonths = new Set();
  for (let month of months) {
    createdMonths.add(month);
  }
  sheet.get().meta().createdMonths = months;
}

async function asSecondClient(func) {
  prefs.loadPrefs();
  prefs.savePrefs({
    groupId: 'group',
    lastSyncedTimestamp: Timestamp.zero.toString(),
  });

  await func();

  await global.emptyDatabase()();
  prefs.savePrefs({
    groupId: 'group',
    lastSyncedTimestamp: Timestamp.zero.toString(),
  });
}

function expectCellToExist(sheetName, name) {
  let value = sheet.get().getCellValueLoose(sheetName, name);
  expect(value).not.toBe(null);
}

function expectCellNotToExist(sheetName, name, voided?: boolean) {
  let value = sheet.get().getCellValueLoose(sheetName, name);
  expect(value).toBe(voided ? 0 : null);
}

describe('Sync projections', () => {
  test('synced categories should have budgets created', async () => {
    let groupId, fooId, barId;
    await asSecondClient(async () => {
      await sheet.loadSpreadsheet(db);
      groupId = await db.insertCategoryGroup({ id: 'group1', name: 'group1' });
      fooId = await db.insertCategory({ name: 'foo', cat_group: 'group1' });
      barId = await db.insertCategory({ name: 'bar', cat_group: 'group1' });
    });

    await sheet.loadSpreadsheet(db);
    registerBudgetMonths(['2017-01', '2017-02']);
    expectCellNotToExist('budget201701', 'sum-amount-' + fooId);
    expectCellNotToExist('budget201701', 'sum-amount-' + barId);
    expectCellNotToExist('budget201701', 'group-sum-amount-' + barId);

    await fullSync();

    // Make sure the budget cells have been created
    expectCellToExist('budget201701', 'sum-amount-' + fooId);
    expectCellToExist('budget201702', 'sum-amount-' + fooId);
    expectCellToExist('budget201701', 'sum-amount-' + barId);
    expectCellToExist('budget201702', 'sum-amount-' + barId);
    expectCellToExist('budget201701', 'group-sum-amount-' + groupId);
    expectCellToExist('budget201702', 'group-sum-amount-' + groupId);
  });

  test('creating and deleting categories in same sync', async () => {
    // It should work when the client creates a category and deletes
    // it in the same sync (should do nothing)
    let fooId;
    await asSecondClient(async () => {
      await sheet.loadSpreadsheet(db);
      await db.insertCategoryGroup({ id: 'group1', name: 'group1' });
      fooId = await db.insertCategory({ name: 'foo', cat_group: 'group1' });
      await db.deleteCategory({ id: fooId });
    });

    await sheet.loadSpreadsheet(db);
    registerBudgetMonths(['2017-01', '2017-02']);
    expectCellNotToExist('budget201701', 'sum-amount-' + fooId);
    await fullSync();
    expectCellNotToExist('budget201701', 'sum-amount-' + fooId);
  });

  test('synced categories should have budgets deleted', async () => {
    let fooId;
    await asSecondClient(async () => {
      await sheet.loadSpreadsheet(db);
      await db.insertCategoryGroup({
        id: 'group1',
        name: 'group1',
      });
      fooId = await db.insertCategory({ name: 'foo', cat_group: 'group1' });
      await db.deleteCategory({ id: fooId });
    });

    await sheet.loadSpreadsheet(db);
    registerBudgetMonths(['2017-01', '2017-02']);

    // Get all the messages. We'll apply them in two passes
    let messages = mockSyncServer.getMessages();

    // Apply all but the last message (which deletes the category)
    await applyMessages(messages.slice(0, -1));
    expect((await db.getCategories()).length).toBe(1);
    expectCellToExist('budget201701', 'sum-amount-' + fooId);

    // Apply the last message and make sure it deleted the appropriate
    // budget cells
    await applyMessages([messages[messages.length - 1]]);
    expect((await db.getCategories()).length).toBe(0);
    expectCellNotToExist('budget201701', 'sum-amount-' + fooId, true);
  });

  test('creating and deleting groups in same sync', async () => {
    // It should work when the client creates a category and deletes
    // it in the same sync (should do nothing)
    let groupId;
    await asSecondClient(async () => {
      await sheet.loadSpreadsheet(db);
      groupId = await db.insertCategoryGroup({ id: 'group1', name: 'group1' });
      await db.deleteCategoryGroup({ id: groupId });
    });

    await sheet.loadSpreadsheet(db);
    registerBudgetMonths(['2017-01', '2017-02']);
    expectCellNotToExist('budget201701', 'group-sum-amount-' + groupId);
    await fullSync();
    expectCellNotToExist('budget201701', 'group-sum-amount-' + groupId);
  });

  test('synced groups should have budgets deleted', async () => {
    // Create the message list which creates categories and groups and
    // then deletes them. Go ahead and include a category deletion in
    // the same pass to make sure that works.
    let groupId, fooId;
    await asSecondClient(async () => {
      await sheet.loadSpreadsheet(db);
      groupId = await db.insertCategoryGroup({
        id: 'group1',
        name: 'group1',
      });
      fooId = await db.insertCategory({ name: 'foo', cat_group: 'group1' });
      await db.deleteCategory({ id: fooId });
      await db.deleteCategoryGroup({ id: groupId });
    });

    await sheet.loadSpreadsheet(db);
    registerBudgetMonths(['2017-01', '2017-02']);

    // Get all the messages. We'll apply them in two passes
    let messages = mockSyncServer.getMessages();

    let firstMessages = messages.filter(m => m.column !== 'tombstone');
    let secondMessages = messages.filter(m => m.column === 'tombstone');

    // Apply all the good messages
    await applyMessages(firstMessages);
    expect((await db.getCategories()).length).toBe(1);
    expect((await db.getCategoriesGrouped()).length).toBe(1);
    expectCellToExist('budget201701', 'sum-amount-' + fooId);
    expectCellToExist('budget201701', 'group-sum-amount-' + groupId);

    // Apply the messages that deletes it
    await applyMessages(secondMessages);
    expect((await db.getCategories()).length).toBe(0);
    expect((await db.getCategoriesGrouped()).length).toBe(0);
    expectCellNotToExist('budget201701', 'sum-amount-' + fooId, true);
    expectCellNotToExist('budget201701', 'group-sum-amount-' + groupId, true);
  });

  test('categories should update the budget when moved', async () => {
    let groupId, fooId;
    await asSecondClient(async () => {
      await sheet.loadSpreadsheet(db);
      groupId = await db.insertCategoryGroup({ id: 'group1', name: 'group1' });
      await db.insertCategoryGroup({ id: 'group2', name: 'group2' });
      fooId = await db.insertCategory({ name: 'foo', cat_group: 'group1' });
      await db.moveCategory(fooId, 'group2');
    });

    await sheet.loadSpreadsheet(db);
    registerBudgetMonths(['2017-01', '2017-02']);

    // Get all the messages. We'll apply them in two passes
    let messages = mockSyncServer.getMessages();

    let firstMessages = messages.slice(0, -2);
    let secondMessages = messages.slice(-2);

    // Apply all the good messages
    await applyMessages(firstMessages);
    let [cat] = await db.getCategories();
    expect(cat.cat_group).toBe('group1');
    expectCellToExist('budget201701', 'group-sum-amount-' + groupId);

    // Apply the messages that deletes it
    await applyMessages(secondMessages);
  });
});
