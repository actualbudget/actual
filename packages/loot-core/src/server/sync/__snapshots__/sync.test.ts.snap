// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sync should send messages to the server 1`] = `
Array [
  Object {
    "column": "amount",
    "dataset": "transactions",
    "id": 1,
    "row": "foo",
    "timestamp": "1970-01-02T10:17:36.789Z-0000-0000000000000id1",
    "value": "N:3200",
  },
  Object {
    "column": "amount",
    "dataset": "transactions",
    "id": 2,
    "row": "foo",
    "timestamp": "1970-01-02T10:17:37.789Z-0000-0000000000000id1",
    "value": "N:4200",
  },
]
`;

exports[`Sync should send messages to the server 2`] = `
Array [
  Object {
    "clock": "{\\"timestamp\\":\\"1970-01-02T10:17:37.789Z-0000-0000000000000id1\\",\\"merkle\\":{\\"2\\":{\\"2\\":{\\"1\\":{\\"1\\":{\\"0\\":{\\"1\\":{\\"2\\":{\\"hash\\":-1992470180},\\"hash\\":-1992470180},\\"hash\\":-1992470180},\\"hash\\":-1992470180},\\"hash\\":-1992470180},\\"hash\\":-1992470180},\\"hash\\":-1992470180},\\"hash\\":-1992470180}}",
    "id": 1,
  },
]
`;
